modules/telnetd/src/main/scala/interact.scala

```diff
@@ -0,0 +1,73 @@
+package gem
+package telnetd
+
+import gem.dao.UserDao
+import gem.enum.ProgramRole
+
+import doobie.imports.{ Capture => DCapture, Transactor }
+import scalaz._, Scalaz._, scalaz.effect._, scalaz.concurrent.Task
+import tuco._, Tuco._
+
+/** Module defining the behavior of our telnet server, parameterized over transactors. */
+object Interaction {
+
+  /**
+   * Initial state for a command shell session has a logged-in Service value, as well as all the
+   * Gem commands and a customized prompt.
+   */
+  def initialState(service: Service[SessionIO]): Session[Service[SessionIO]] =
+    Session.initial(service).copy(
+      commands = Builtins[Service[SessionIO]] |+| command.All,
+      prompt   = s"${service.user.id}@gem> "
+    )
+
+  /**
+   * Login loop. Prompt for a password `remainingTries` times and return a Service on success. The
+   * transactors are needed for the `Service`.
+   */
+  def loginLoop(
+    user:     String,
+    maxTries: Int,
+    xa:       Transactor[SessionIO],
+    txa:      Transactor[Task]
+  ): SessionIO[Option[Service[SessionIO]]] = {
+    def go(remaining: Int): SessionIO[Option[Service[SessionIO]]] =
```

Comment:
The login loop has telnet effects, so it's in `SessionIO`. 

If the login loop is successful it will yield a `Service` object, which provides service calls in an arbitrary IO-like effect type (could be `Task` for instance, and probably will be for the web front end). Here we're doing service calls from `SessionIO` in our command shell so that's the convenient effect type to use.